pkg/kor/namespaces.go

```diff
@@ -0,0 +1,221 @@
+package kor
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	types "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
+
+	"github.com/yonahd/kor/pkg/filters"
+)
+
+type GenericResource struct {
+	NamespacedName types.NamespacedName
+	GVR            schema.GroupVersionResource
+}
+
+func getGVR(name string, splitGV []string) (*schema.GroupVersionResource, error) {
+	switch NumberOfGVPartsFound := len(splitGV); NumberOfGVPartsFound {
+	case 1:
+		return &schema.GroupVersionResource{
+			Version:  splitGV[0],
+			Resource: name,
+		}, nil
+	case 2:
+		return &schema.GroupVersionResource{
+			Group:    splitGV[0],
+			Version:  splitGV[1],
+			Resource: name,
+		}, nil
+	default:
+		return nil, fmt.Errorf("gv is wrong length slice: %d", NumberOfGVPartsFound)
+	}
+}
+
+func ignoreResourceType(resource string, ignoreResourceTypes []string) bool {
+	for _, ignoreType := range ignoreResourceTypes {
+		if resource == ignoreType {
+			return true
+		}
+	}
+	return false
+}
+
+func getNamespacedResources(
+	ctx context.Context,
+	clientset kubernetes.Interface,
+	dynamicClient dynamic.Interface,
+	namespace string,
+	filterOpts *filters.Options,
+) (*[]GenericResource, error) {
+	var genResources []GenericResource
+
+	apiResourceLists, err := clientset.Discovery().ServerPreferredNamespacedResources()
+	if err != nil {
+		panic(err)
+	}
+
+	// Iterate over all API resources and list instances of each in the specified namespace
+	for _, apiResourceList := range apiResourceLists {
+		for _, apiResource := range apiResourceList.APIResources {
+			gv := strings.Split(apiResourceList.GroupVersion, "/")
+			gvr, err := getGVR(apiResource.Name, gv)
+			if err != nil {
+				return nil, err
+			}
+
+			unstructuredList, err := dynamicClient.Resource(*gvr).Namespace(namespace).List(ctx, metav1.ListOptions{})
+			if err != nil {
+				continue
+			}
+
+			for _, unstructuredObj := range unstructuredList.Items {
+				gr := GenericResource{
+					GVR: *gvr,
+					NamespacedName: types.NamespacedName{
+						Namespace: unstructuredObj.GetNamespace(),
+						Name:      unstructuredObj.GetName(),
+					},
+				}
+				// Specific list of resources to ignore - resources created in all namespaced by default
+				if gr.GVR.Resource == "configmaps" && gr.GVR.Version == "v1" && gr.NamespacedName.Name == "kube-root-ca.crt" {
+					continue
+				}
+				if gr.GVR.Resource == "serviceaccounts" && gr.GVR.Version == "v1" && gr.NamespacedName.Name == "default" {
+					continue
+				}
+				if gr.GVR.Resource == "events" {
+					continue
+				}
+				// User specified resource type ignore list
+				if ignoreResourceType(gr.GVR.Resource, filterOpts.IgnoreResourceTypes) {
+					continue
+				}
+				genResources = append(genResources, gr)
```

Comment:
Why are we collecting and not returning false?

Reply:
This one is to ignore user specified resource types, for example we use cluster level externalsecrets, which are creating secrets in each namespace, where we consider namespace to be empty if there are only secrets left in the namespace. One can use:
```
kor ns --ignore-resource-types secrets,configmaps
```
that should list any namepsaces where only cms and secrets are left (default resources and events are ignored as well).

Reply:
From what I see we check if this list is longer than 0 to determine if the namespace is unused. Why continue collecting these or iterating after we find one?

Reply:
true, I'll rework this function to return true/false - as an answer to "resources found?" . This should speed things up significantly.

Reply:
@yonahd fixed this